### ISSUE TITLE
Hash value in instance variable does not accept subtypes

### ISSUE THREAD

**Issue by TPei** [crystal-lang/crystal#6728]
When defining a `Hash` with values of `SomeClass` it is perfectly fine to have values be of type `SomeSubClass < SomeClass` as long as the hash is not an instance variable, then it fails with
`instance variable '@mapping' of SomeClass must be Hash(String, SomeThing), not Hash(String, SomeSubThing)`

```
class SomeClass
  def some_method(mapping = {} of String => SomeThing)
    mapping["test"].bla
  end
end

=> works as expected
```

```
class SomeClass
  @mapping = {} of String => SomeThing
  
  def some_method(@mapping = {} of String => SomeThing)
    @mapping["test"].bla
  end
end

=> does not work
```

See https://play.crystal-lang.org/#/r/506z and https://play.crystal-lang.org/#/r/5070 for more extensive comparison

**Comment by asterite** [comment]
Please check https://github.com/crystal-lang/crystal/issues/3803

You have to do `{ "test" => SomeSubThing.new } of String => SomeClass` for that to work.

tl;dr:
- the compiler infers the most strict type for an array/hash literal
- restrictions let subclasses pass
- instance variables don't allow subclasses (yes, as long as the compile-time type is the same)

**Comment by asterite** [comment]
This is one of the most confusing aspects of the language. Unfortunately we don't know how to improve it (or nobody put a lot of time into thinking about this).

**Comment by TPei** [comment]
ah, nice, thanks for the info!
